community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/ExternalPropertiesDecoratorIT.java

```diff
@@ -0,0 +1,145 @@
+import org.neo4j.unsafe.impl.batchimport.input.InputNode;
+import org.neo4j.unsafe.impl.batchimport.input.UpdateBehaviour;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+
+import static java.lang.String.valueOf;
+
+import static org.neo4j.csv.reader.Readables.wrap;
+import static org.neo4j.unsafe.impl.batchimport.input.InputEntityDecorators.NO_NODE_DECORATOR;
+import static org.neo4j.unsafe.impl.batchimport.input.csv.DataFactories.data;
+import static org.neo4j.unsafe.impl.batchimport.input.csv.DataFactories.defaultFormatNodeFileHeader;
+import static org.neo4j.unsafe.impl.batchimport.input.csv.DataFactories.nodeData;
+
+public class ExternalPropertiesDecoratorIT
+{
+    @Test
+    public void shouldDecorateExternalPropertiesInParallelProcessingCsvInput() throws Exception
+    {
+        // GIVEN
+        int processors = 5;
+        Collector collector = mock( Collector.class );
+        int count = 1000;
+        Configuration config = new Configuration.Overriden( Configuration.COMMAS )
+        {
+            @Override
+            public int bufferSize()
+            {
+                // Keep this low so that there will be many batches, to exercise the parallel processing
+                // 300 is empirically measured to roughly produce ~20 chunks
+                return 300;
+            }
+        };
+        IdType idType = IdType.STRING;
+        Decorator<InputNode> decorator = new ExternalPropertiesDecorator(
+                data( NO_NODE_DECORATOR, () -> decoratedData( count ) ),
+                defaultFormatNodeFileHeader(),
+                config, idType, UpdateBehaviour.ADD, collector );
+        Input input = new CsvInput(
+                nodeData( data( decorator, () -> mainData( count ) ) ), defaultFormatNodeFileHeader(),
+                null, null,
+                idType, config,
+                collector, processors );
+
+        // WHEN/THEN
+        try ( InputIterator<InputNode> nodes = input.nodes().iterator() )
+        {
+            int i = 0;
+            for ( ; i < count; i++ )
+            {
+                assertTrue( nodes.hasNext() );
+                InputNode node = nodes.next();
+                // This property comes from decorator
+                assertHasProperty( node, "extra", node.id() + "-decorated" );
+                if ( i == 0 )
+                {
+                    nodes.processors( processors - nodes.processors( 0 ) );
+                }
+            }
+            assertEquals( count, i );
+            assertFalse( nodes.hasNext() );
+        }
+    }
+
+    private void assertHasProperty( InputNode node, String key, Object value )
+    {
+        Object[] properties = node.properties();
+        boolean found = false;
+        for ( int i = 0; i < properties.length; i++ )
+        {
+            if ( properties[i++].toString().equals( key ) )
```

Comment:
Document how properties are stored in `InputEntity` (`[key|value|key|value...]`)